File: _4_Longest_Common_Prefix.ts
function longestCommonPrefix(strs: string[]): string {
    if (strs.length === 0) return "";

    let pref = strs[0];
    let prefixLen = pref.length;

    for (let i = 1; i < strs.length; i++) {
        const word = strs[i];

        while (prefixLen > word.length || pref !== word.substring(0, prefixLen)) {
            prefixLen--;
            if (prefixLen === 0) return "";
            pref = pref.substring(0, prefixLen);
        }
    }

    return pref;
}

// Ejemplos de uso:
const strs1 = ["flower","flow","flight"];
console.log("Result:", longestCommonPrefix(strs1)); // "fl"

const strs2 = ["dog","racecar","car"];
console.log("Result:", longestCommonPrefix(strs2)); // ""

File: _4_Longest_Common_Prefix.java
import java.lang.StringBuilder;

public class _4_Longest_Common_Prefix {
    public static void main(String[] args) {
        Solution sol = new Solution();
        String [] strs = {"flower","flow","flight"};
        String result = sol.longestCommonPrefix(strs);
        System.out.println("Result: " + result);
        String [] strs2 = {"dog","racecar","car"};
        result = sol.longestCommonPrefix(strs2);
        System.out.println("Result: " + result);

    }
}

//My solution
class Solution {
    public String longestCommonPrefix(String[] strs) {

        String shortCad = strs[0];

        for (String s : strs) {
            if (s.length() < shortCad.length()) {
                shortCad = s;
            }
        }

        StringBuilder ans = new StringBuilder();

        for (int i = 0; i < shortCad.length(); i++) {
            char c = shortCad.charAt(i);
            for (int j = 0; j < strs.length; j++) {
                char d = strs[j].charAt(i);
                if (c != d) {
                    return ans.toString();
                }
            }
            ans.append(c);
        }
        return ans.toString();
    }
}